### DISCUSSION TITLE
Confusion

### DISCUSSION THREAD

**Discussion by alexchudnovsky** [pmndrs/zustand#778] [Q&A] [answered]
Hi there,

I have a small issue that I can't understand, wondering if someone can point me in the right direction

Please see the snippet below

Could someone please explain why **console.log("The user is", user)** is returning null ?

as loadUserFromAsyncStorage() - is using await 

Shouldn't the console.log return a user object instead of null ?

Thank you so much if someone can clear this up

---

    const { user, loadUserFromAsyncStorage} = userState();

    useEffect(() => {

        loadUserAndPrepareMessaging()

    }, [])

---

    async function loadUserAndPrepareMessaging() {

        try {

            await loadUserFromAsyncStorage()
            **console.log("The user is", user)**  -- this returns null
            await requestUserPermission()

        } catch (error) {

            console.log("Error loading user from storage", error.message)

        } finally {
            setLoading(false)
        }

    }

----

export const userState = create((set, get) => ({

    user: null,
 

    loadUserFromAsyncStorage: async () => {

        if (!get().user)

            try {
                let storedUser = JSON.parse(await AsyncStorage.getItem(API.asyncStorageKey))

                console.log("The stored user is", storedUser)

                //we don't care about the data which is returned, we just want to ensure that the jwt token is still valid

                await axios.get(`${API.server}/users/me`, {

                    headers: {
                        Authorization:
                            `Bearer ${storedUser.jwt}`,
                    }

                })

                set({ user: storedUser })

            } catch (e) {
                console.log("Could not load user data from active storage either because he does not exist or the jwt token is invalid", e.message)
            }
    },

----

**Comment by dai-shi** [comment]
Hi, the code doesn't seem complete. Where does `user` come from?
Would you be able to create a minimal reproduction with codesandbox?

**Comment by alexchudnovsky** [comment]
Please see below  

export const userState = create((set, get) => ({

user: null,


loadUserFromAsyncStorage: async () => {

    if (!get().user)

        try {
            let storedUser = JSON.parse(await AsyncStorage.getItem(API.asyncStorageKey))

            console.log("The stored user is", storedUser)

            //we don't care about the data which is returned, we just want to ensure that the jwt token is still valid

            await axios.get(`${API.server}/users/me`, {

                headers: {
                    Authorization:
                        `Bearer ${storedUser.jwt}`,
                }

            })

            set({ user: storedUser })

        } catch (e) {
            console.log("Could not load user data from active storage either because he does not exist or the jwt token is invalid", e.message)
        }
},

**Reply by alexchudnovsky** [reply]
sorry the formatting is a bit messed up... 

This is the missing part above 

export const userState = create((set, get) => ({

user: null,

loadUserFromAsyncStorage: async () => { ....

**Reply by dai-shi** [reply]
Something like this?
```js
export const userState = create((set, get) => ({
  user: null,
  loadUserFromAsyncStorage: async () => {
    if (!get().user) {
      try {
        let storedUser = JSON.parse(await AsyncStorage.getItem(API.asyncStorageKey))
        console.log("The stored user is", storedUser)
        // we don't care about the data which is returned
        await Promise.resolve("dummy")
        set({ user: storedUser })
      } catch (e) {
        console.log(e.message)
      }
    }
  },
})
```

**Comment by alexchudnovsky** [comment]
Yeah that's exactly right :) thanks

**Reply by alexchudnovsky** [reply]
awww sorry, you have a fix there.. i thought you just formatted it

so the problem is, I am not resolving the promise ?

I should be doing this - await Promise.resolve("dummy") right ?

**Reply by dai-shi** [reply]
No, I didn't mean to fix. It's just simplified as it's said we don't care the data.

**Reply by alexchudnovsky** [reply]
Ahh right.. sorry :)

in that case, yeah, what you have is exactly right.. 

I am thinking, could the issue be caused, because loadUserFromAsyncStorage method is not returning anything ?

**Reply by dai-shi** [reply]
I still don't understand the issue. What's your issue?

**Reply by alexchudnovsky** [reply]
In another area of the application in a useEffect method I am doing the below


const { user, loadUserFromAsyncStorage} = userState();

useEffect(() => {

    try {

        await loadUserFromAsyncStorage()
        **console.log("The user is", user)**  -- this returns null
        await requestUserPermission()

    } catch (error) {

        console.log("Error loading user from storage", error.message)

    } finally {
        setLoading(false)
    }

}

}, [])


But immediately after i call  loadUserFromAsyncStorage() which sets the User even prints it out to the console.

i thin console log the user 

**console.log("The user is", user)**  

But the user returns null ( the user has already been set though )

**Reply by alexchudnovsky** [reply]
thin = then

I just don't understand, why its console logging as null, after the user is set in the  await loadUserFromAsyncStorage()

**Reply by dai-shi** [reply]
You can edit your message. You can also do syntax highlighting.
See this: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting

**Comment by alexchudnovsky** [comment]
Just one question please 

        await Promise.resolve("dummy")
        set({ user: storedUser })

how come the dummy promise is returned before the set method is called ? - or it makes no difference ? .. as long as I return a promise is the important part yeah  ?

**Comment by alexchudnovsky** [comment, marked as answer]
I have understood, the reason for this.. 

It is inside a nested method which is a closure 

Hence its accessing stale closure data

**Reply by dai-shi** [reply]
Great you found the reason.
